### DISCUSSION TITLE
Rebuild equuleus with a patch to accel-ppp

### DISCUSSION THREAD

**Discussion by marekm72** [dd010101/vyos-jenkins#54] [General]
I'm still running official 1.3.8 and would like to test just a single bugfix in accel-ppp without any other changes that could break my proven working production setup, so I need to be careful.

However, the way accel-ppp is built has changed between equuleus and sagitta.  In sagitta there is the accel-ppp patches subdirectory with one patch and it should be easy to add a second one there.  But in equuleus it seems it might need some Jenkins magic I'm not yet too familiar with, and I'd like to get it right the first time.  The nature of the bug is such that I can only test it well under real heavy production use, on live customers who have all sorts of buggy SOHO routers as PPPoE clients - I'm running two core routers at two different locations doing all of BGP+OSPF+PPPoE, so I plan to update the first one at night during one weekend, wait a week or two, and then on another weekend at night update the second one.   If all works well I may try to upgrade patched equuleus to patched sagitta, but I'm not rushing to do it yet as it's still x.0 release and asking politely on their forum about when 1.4.1 is planned seems to not be too welcome.

The small change is here - https://github.com/xebd/accel-ppp/pull/186 - and these two modified files in accel-ppp haven't changed much in years so it should apply cleanly to any version.  What would be the best way to go about it?

If tested and proven to work well, it should apply to equuleus and sagitta (and circinus too), VyOS(tm) is free to include it too of course, but if they don't then perhaps we could carry it in this unofficial fork?  (Not yet, as I want to test it properly first for some time to be reasonably sure it causes no regressions.)

**Comment by dd010101** [comment]
This saggita has method to include external patches but equuleus doesn't so that's a limitation and something that would need to be implemented for long term usage in the build but then you need to include the patch file and that's not so easy when you just want to try it. There is more straightforward way how build custom sources temporarily.

The accel-ppp is build by the `linux-kernel` package, specifically it has it's own shell script `build-accel-ppp.sh`. The source is pulled from `https://github.com/accel-ppp/accel-ppp.git` in the `linux-kernel/Jenkinsfile` thus you need to modify the Jenkinsfile so it uses already patched accel-ppp sources from your own repository. Thus you don't need to deal with the patches since the source is already patched in such case.

This is the snippet of the Jenkinsfile you care about:

```
stage('Accel-PPP') {
    when {
        beforeOptions true
        beforeAgent true
        anyOf {
            changeset pattern: "**/packages/linux-kernel/*"
            changeset pattern: "**/data/defaults.json"
            triggeredBy cause: "UserIdCause"
        }
    }
    steps {
        dir(env.BASE_DIR + '/accel-ppp') {
            checkout([$class: 'GitSCM',
                doGenerateSubmoduleConfigurations: false,
                extensions: [[$class: 'CleanCheckout']],
                branches: [[name: '0b4ef9862c65bf' ]],
                userRemoteConfigs: [[credentialsId: 'GitHub-vyosbot', url: 'https://github.com/accel-ppp/accel-ppp.git']]])
        }
    }
}
```

You need to rewrite the URL to your GIT repository.

You also need to modify the branches name, since the build is expecting the 0b4ef9862c65bf commit you should make branch of accel-ppp at this commit and apply the patch to this commit, then commit/push and then you can use the newly created branch name as replacement for the original commit (0b4ef9862c65bf).

If you can prepare the GIT repo with patched branch based on 0b4ef9862c65bf then you can simply fill in your details into the Jenkinsfile in the accel-ppp section.

You can modify Jenkinsfile in two ways - first you can make fork of vyos-build and use your own vyos-build repo in the Jenkins Pipeline definition, likely too much work. There is much easier way - the Jenkins gives you a option to Replay under given run #number. Thus start build of linux-kernel/equuleus (you can cancel the run, you don't need to wait for it to complete), select the cancelled run and in left panel you have Replay option. This gives you option to edit the Jenkinsfile right from the Jenkins UI and thus you only need the patched accel-ppp repo to make custom build. The Replay launches modified build and then you just wait and eventually you get the result in your reprepro repository. 

This requires bit of GIT usage and some Jenkins interaction but it's the most straightforward way how to build custom changes for test purposes. This also has the advantage that if you make further changes you just push another commit to your repo/branch and hit build - so you can iterate without any additional work.

I didn't actually try this with accel-ppp but it's the general idea I would apply for this type of task.

**Reply by dd010101** [reply]
Since you are using official ISO and you need to test in production environment then it would be bad idea to switch to different ISO to test such changes since then you don't know if you are testing the change you made or the different ISO, although the packages are from the same source and all of that it's generally bad idea to test multiple things all at once, since then you don't know what are you testing exactly... This applies even if you would switch to different official ISO - you should change literally only the thing you are changing if possible and swapping to any different ISO is changing absolutely more than that.

That's why I would prefer to use the identical ISO/image you are using and just apply the one change you want. I didn't try this with accel-ppp but it should work - to wget the accel-ppp_1.12.0-170-*_amd64.deb from reprepro repository and then dpkg -i/reboot and you should have the change applied to existing system without unrelated changes - it's for sure strongly proffered to test changes in isolated fashion where everything else stays the same. If it's possible... You should install another 1.3.8 image, so you have 1.3.8 modified and 1.3.8 stock.

In such case you should also check if your system is using the accep-ppp-1.12.0-170 version, I have no idea what VyOS 1.3.8 is using. Maybe the 0b4ef9862c65bf commit isn't the correct one - this is taken from current sources. Thus if you have different version you may want to branch on different commit.

**Reply by marekm72** [reply]
Thanks, installing just that one package or even only the single file /usr/sbin/accel-pppd looks like the best way.  Original 1.3.8 has accel-ppp-1.12.0-170-g0b4ef98 so it matches the commit.  The build container could change since the 1.3.8 release time, hopefully not too much - there might be 1.3.9 later, who knows...

**Reply by dd010101** [reply]
The vyos-build currently produces `accel-ppp_1.12.0-170-g0b4ef98_amd64.deb` and this matches what did you post and also the describe in the version g**0b4ef98** is the [0b4ef9862c65bf](https://github.com/accel-ppp/accel-ppp/commit/0b4ef9862c65bf) commit. Thus if you base your accel-ppp changes on this commit you should get the same thing.

**Reply by dd010101** [reply]
Thus something like this should get you get the same accel-ppp but with your patch:
```
git clone https://github.com/accel-ppp/accel-ppp.git
cd accel-ppp

git checkout -b patch 0b4ef9862c65bf
wget https://patch-diff.githubusercontent.com/raw/xebd/accel-ppp/pull/186.patch -O ../pull-request.patch
git am ../pull-request.patch

git remote add fork git@github.com:OWNER/REPOSITORY.git
git push fork patch
```

Then you can use the `https://github.com/OWNER/REPOSITORY.git` together with branch name `patch` in the **Jenkins Replay**.

Original:
```
dir(env.BASE_DIR + '/accel-ppp') {
	checkout([$class: 'GitSCM',
		doGenerateSubmoduleConfigurations: false,
		extensions: [[$class: 'CleanCheckout']],
		branches: [[name: '0b4ef9862c65bf' ]],
		userRemoteConfigs: [[credentialsId: 'GitHub-vyosbot', url: 'https://github.com/accel-ppp/accel-ppp.git']]])
}
```

Modified:
```
dir(env.BASE_DIR + '/accel-ppp') {
	checkout([$class: 'GitSCM',
		doGenerateSubmoduleConfigurations: false,
		extensions: [[$class: 'CleanCheckout']],
		branches: [[name: 'patch' ]],
		userRemoteConfigs: [[credentialsId: 'GitHub-vyosbot', url: 'https://github.com/OWNER/REPOSITORY.git']]])
}
```

Of course replace the `OWNER/REPOSITORY` with yours details.

This should yield `accel-ppp_1.12.0-172-gHASH_amd64.deb` since your patch contains two new commits.

**Reply by marekm72** [reply]
I have tried to edit Jenkinsfile and Replay - it worked the first time and made a .deb package from which I extracted just the accel-pppd binary, and replaced the original one in the running system after making a backup.  But it doesn't yet work properly and I need to try again with some changes https://github.com/marekm72/accel-ppp/tree/patch - but after the second Replay the build failed, the log is huge.  Iterations take a long time as it builds all of the kernel each time (because accel-ppp includes some kernel modules), not sure how to run just the accel-ppp build manually in the correct environment (container with old Debian for equuleus).

**Reply by dd010101** [reply]
The accel-ppp depends on the kernel thus you can't just build accel-ppp alone without pointing the build to the correct kernel headers and since VyOS is using it's own kernel you can't just use whatever Debian nor the container since that would build the kernel module for the kernel of the Debian or the container you are using thus not for VyOS kernel. That's why it's included as one package with kernel and firmware and that's why the build is huge and takes long time.

So how do you build accel-ppp alone? It would be easiest to use the vyos-build docker container of equuleus as the build environment as the Jenkins does it. Thus everything should be executed inside the docker container and thus it's irrelevant what is your host OS/version.

Firstly you need the kernel headers - they are built together with the kernel, thus you need to build kernel first and then obtain the `apt/equuleus/pool/main/l/linux-5.4.268-amd64-vyos/linux-headers-5.4.268-amd64-vyos_5.4.268-1_amd64.deb` or equivalent for your kernel. You can extract this or dpkg -i, doesn't really matter, you just need a whatever path where the headers are.

You can abuse the build script `build-accel-ppp.sh`. But that won't work alone, since it expects $(pwd)/kernel-vars and this file is generated by the [build-kernel.sh](https://github.com/vyos/vyos-build/blob/equuleus/packages/linux-kernel/build-kernel.sh#L37). Thus you need to create this file with the correct variables pointing to the directory where you have the headers and also the version/suffix of the kernel.

Then in theory the script should yield accel-ppp.deb alone.

**Reply by dd010101** [reply]
That's wrong, there is bigger dependency on the kernel and it seems you need to actually build it after all? I'm not quite sure what is exactly required but if you don't build kernel then accel-ppp doesn't build and if you do build the kernel then accel-ppp builds too, do you need some kind of configuration or what? I don't know but it's easier to just build the kernel one time and you don't need to think about it. I have no idea how building of kernel drivers works but if you follow what the Jenkins does then it simply works.

You can skip building the kernel again and again, you can do it once and then build/rebuild accel-ppp alone once you have the environment prepared. The Jenkins builds everything every time since the Jenkins doesn't know if something changed in the meantime that's why there is no persistent state.

Something like this works standalone:
```bash
# get the scripts
git clone https://github.com/dd010101/vyos-build
cd vyos-build
git checkout equuleus

# run inside vyos-build container of equuleus
docker run --rm --name accel-build -it --privileged -v $(pwd):/vyos -w /vyos vyos/vyos-build:equuleus bash

# build kernel once
cd /vyos/packages/linux-kernel
KERNEL_VER=$(cat /vyos/data/defaults.json | jq -r .kernel_version)
curl -OL https://www.kernel.org/pub/linux/kernel/v5.x/linux-${KERNEL_VER}.tar.xz
tar xf linux-${KERNEL_VER}.tar.xz
ln -s linux-${KERNEL_VER} linux
./build-kernel.sh

# now you can build and rebuild the accel-ppp
git clone https://github.com/accel-ppp/accel-ppp.git
cd accel-ppp
git checkout 0b4ef9862c65bf
cd ..

./build-accel-ppp.sh
```

This yields `accel-ppp_1.12.0-170-g0b4ef98_amd64.deb` in current directory (in the container and also on the host).

More complicated then I would guess at first. Some things you just need to try... The accel-ppp is not regular package, it provides extra drivers and that's why it can't be simply compiled without the kernel but I would guess the source code of the kernel (or the headers) should be enough but the driver compilation failed even if I configured the kernel sources, so I'm missing some step that's included in the complication and that's why it only works after you compile the kernel I guess. Maybe I just had wrong path to the headers, who knows... Normally for drivers the headers are enough...

**Reply by marekm72** [reply]
Thanks, exactly what I was looking for - can build the kernel just once, then make changes and repeat accel-ppp builds as many times as needed.  The thing really shoud be spilt in separate packages, accel-ppp itself (the PPPoE server, I don't use the PPTP/L2TP/SSTP/IPoE parts) is not dependent on the drivers except kernel-mode PPPoE which is already in the standard kernel.
